src/libsyntax/feature_gate.rs

```diff
@@ -1507,6 +1507,42 @@ impl<'a> PostExpansionVisitor<'a> {
 }
 
 impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
+    fn visit_ident(&mut self, ident: Ident) {
+        // `await` may become a keyword in the 2018 edition and later,
+        // but it is needed for nightly experimentation with the
+        // async_await feature, so we can't make it a proper keyword,
+        // nor can we make it a keyword based on a feature gate since
+        // features aren't available during parsing. So, we produce
+        // a custom error if `await` is used as an identifier on the
+        // 2018 edition or later and the `async_await` feature isn't
+        // enabled.
+        if ident.name == "await"
```

Comment:
I have a vague fear that this may not work for some uses, because it runs post-expansion.

For example, what if I have (in a Rust 2018 crate):

```rust
macro_rules! r#await {
    () => { println!("Hello, world!") }
}

fn main() { await!(); }
```

Does that error?

Reply:
This is why I was considering saying that we should adapt the keyword warning lint we use for migrating to give a hard error in this case.

Reply:
Originally I made the code in the lint issue an error, but then I discovered that issuing an error (by setting `diagnostic.level = Level::Error`) from the context of a lint won't give you any error output if the compilation fails unless the lint is enabled-- it will just print out "error: ". You're right, though, that that example might not work, so I'll try it out and see if I can hack around the diagnostic builder in the lint code.

Reply:
Heh, actually, your example does error-- but not where it should:

```
error[E0721]: `await` is a keyword in the 2018 edition                                                                                              
  --> $DIR/async_keyword_again.rs:3:14                                                                                                              
   |                                                                                                                                                
LL | macro_rules! r#await {                                                                                                                         
   |              ^^^^^^^ help: you can use a raw identifier to stay compatible: `r#await`                                                          
   |                                                                                                                                                
   = help: if you were trying to use async/await!, add #![feature(async_await)] to the crate attributes to enable                                                                                                                                                                                       
error: aborting due to previous error                                                                                                               
                                                                                                                                                    
For more information about this error, try `rustc --explain E0721`.  
```
I'll look into changing the lint.

Reply:
Done.